src/sbvr-api/permissions.ts

```diff
@@ -526,31 +527,74 @@ const generateConstrainedAbstractSql = (
 	// permissions circles.
 	const canAccessTrace: string[] = [resourceName];
 
-	const canAccessFunction: ResourceFunction = function (property: AnyObject) {
+	const resolveBind = (maybeBind: any, extraBinds: ODataBinds) => {
+		if (isBindReference(maybeBind)) {
+			const { bind } = maybeBind;
+			if (typeof bind === 'string' || bind < odata.binds.length) {
+				return odata.binds[bind];
+			}
+			return extraBinds[bind - odata.binds.length];
+		}
+		return maybeBind;
+	};
+	const canAccessFunction: ResourceFunction = function (property) {
 		// remove method property so that we won't loop back here again at this point
-		delete property.method;
+		const { method, ...resolvedProperty } = property;
 
 		if (!this.defaultResource) {
 			throw new Error(`No resource selected in AST.`);
 		}
 
 		const targetResource = this.NavigateResources(
 			this.defaultResource,
-			property.name,
+			resolvedProperty.name,
 		);
 
+		const lambdaId = `${lambdaAlias}+${inc}`;
+		inc = inc + 1;
+
 		const targetResourceName = sqlNameToODataName(targetResource.resource.name);
 
-		if (canAccessTrace.includes(targetResourceName)) {
-			// we don't want to allow permission loops for now, therefore we are
-			// throwing the exception here. If we ever want to allow permission
-			// loops return a false AST statement here (like true eq false), to
-			// not recursivley follow query branches in a deep first search.
-			throw new PermissionError(
-				`Permissions for ${resourceName} form a circle by the following path: ${canAccessTrace.join(
-					' -> ',
-				)} -> ${targetResourceName}`,
-			);
+		const traceIndex = canAccessTrace.findIndex(
+			(rName) => rName === targetResourceName,
+		);
+		if (traceIndex !== -1) {
+			if (canAccessTrace[canAccessTrace.length - 1] !== targetResourceName) {
+				throw new Error(
+					`Indirectly circular 'canAccess()' permissions are not supported, currently permissions for ${resourceName} form an indirect circle by the following path: ${canAccessTrace.join(
+						' -> ',
+					)} -> ${targetResourceName}`,
+				);
+			}
+
+			const { args } = method[1];
```

Comment:
how do you know that entry at idx 1 exists?

Reply:
Because the type is `[ 'call', { args: any[], method: string } ]` and is checked before we ever enter this function

Reply:
thx. helps me understand the code better 👍